File: Golang/DesignPatterns/StateDesignPattern.go
package main

import "fmt"

type IATMMachineState interface {
	InsertCard()
	WithdrawCash()
	RemoveCard()
	PressCancel()
}

// ATM: CardInsertedState, NoMoneyState, HasMoneyState, NoCardState

type CardInsertedState struct {
	atm *ATMMachine
}

func (s *CardInsertedState) InsertCard() {
	fmt.Println("Already Inserted No use")
}

func (s *CardInsertedState) WithdrawCash() {
	fmt.Println("Some 100 Line  of code you are running")
}

func (s *CardInsertedState) RemoveCard() {
	fmt.Println("Card is removed")
	s.atm.SetState(s.atm.GetNoCardState()) // logic for transitioning the state
}

func (s *CardInsertedState) PressCancel() {
	s.RemoveCard()
}

type NoCardInsertedState struct {
	atm *ATMMachine
}

func (s *NoCardInsertedState) InsertCard()   {}
func (s *NoCardInsertedState) WithdrawCash() {}
func (s *NoCardInsertedState) RemoveCard()   {}
func (s *NoCardInsertedState) PressCancel()  {}

type HasMoneyState struct {
	atm *ATMMachine
}

func (s *HasMoneyState) InsertCard()   {}
func (s *HasMoneyState) WithdrawCash() {}
func (s *HasMoneyState) RemoveCard()   {}
func (s *HasMoneyState) PressCancel()  {}

type ATMMachine struct {
	currentState        IATMMachineState
	cardInsertedState   IATMMachineState
	noCardInsertedState IATMMachineState
	hasMoneyState       IATMMachineState
}

func NewATMMachine() *ATMMachine {
	atm := &ATMMachine{}
	atm.noCardInsertedState = &NoCardInsertedState{atm: atm}
	atm.cardInsertedState = &CardInsertedState{atm: atm}
	atm.hasMoneyState = &HasMoneyState{atm: atm}
	atm.currentState = atm.noCardInsertedState
	return atm
}

func (a *ATMMachine) SetState(state IATMMachineState) {
	a.currentState = state
}

func (a *ATMMachine) GetNoCardState() IATMMachineState {
	return a.noCardInsertedState
}

func (a *ATMMachine) GetCardInsertedState() IATMMachineState {
	return a.noCardInsertedState
}

func (a *ATMMachine) GetHasMoneyState() IATMMachineState {
	return a.noCardInsertedState
}

func (a *ATMMachine) InsertCard()   { a.currentState.InsertCard() }
func (a *ATMMachine) WithdrawCash() { a.currentState.WithdrawCash() }
func (a *ATMMachine) RemoveCard()   { a.currentState.RemoveCard() }
func (a *ATMMachine) PressCancel()  { a.currentState.PressCancel() }

func main() {}

File: Java/DesignPatterns/StateDesignPattern.java
package DesignPatterns;


interface IATMMachineState{
    void insertCard();
    void withdrawCash();
    void removeCard();
    void pressCancel();
}



//it will have some states 

// ATM: CardInsertedState, NoMoneyState, HasMoneyState, NoCardState

class CardInsertedState implements IATMMachineState{
    private ATMMachine atm;
    
    public CardInsertedState(ATMMachine atm){
        this.atm = atm;
    }
    @Override
    public void insertCard() {
        System.out.println("Already Inserted No use");
    }

    @Override
    public void withdrawCash() {
        System.out.println("Some 100 Line  of code you are running");
    }

    @Override
    public void removeCard() {
        System.out.println("Card is removed");
        atm.setState(atm.getNoCardState()); //logic for transitioning the state
    }

    @Override
    public void pressCancel() {
        removeCard();
    }
}


class No_CardInsertedState implements IATMMachineState{
    private ATMMachine atm;
    public No_CardInsertedState(ATMMachine atm){
        this.atm = atm;
    }
    @Override
    public void insertCard() {

    }

    @Override
    public void withdrawCash() {

    }

    @Override
    public void removeCard() {

    }

    @Override
    public void pressCancel() {

    }
}


class HasMoneyState implements IATMMachineState{
    private ATMMachine atm;
    public HasMoneyState(ATMMachine atm){
        this.atm = atm;
    }
    @Override
    public void insertCard() {

    }

    @Override
    public void withdrawCash() {

    }

    @Override
    public void removeCard() {

    }

    @Override
    public void pressCancel() {

    }
}


class ATMMachine{
    IATMMachineState currentState;
    IATMMachineState cardInsertedState;
    IATMMachineState no_CardInsertedState;
    IATMMachineState hasMoneyState;
    
    public ATMMachine(){
        currentState = new No_CardInsertedState(this);
        cardInsertedState = new CardInsertedState(this);
        no_CardInsertedState = new No_CardInsertedState(this);
        hasMoneyState = new HasMoneyState(this);
    }

    void setState(IATMMachineState state){
        this.currentState = state;
    }

    IATMMachineState getNoCardState(){
        return no_CardInsertedState;
    }
    IATMMachineState getCardInsertedState(){
        return no_CardInsertedState;
    }

    IATMMachineState getHasMoneyState(){
        return no_CardInsertedState;
    }


    void insertCard(){
        currentState.insertCard();
    }
    void withdrawCash(){
        currentState.withdrawCash();
    }
    void removeCard(){
        currentState.removeCard();
    }
    void pressCancel(){
        currentState.pressCancel();
    }


}


public class StateDesignPattern {
    
}
